lib/spack/spack/test/cmd/env.py

```diff
@@ -1921,6 +1933,136 @@ def test_env_include_concrete_remove_env():
+    # environment with "concretizer:reuse:true" the included
+    # concrete spec overrides the default with mpi@1.0.
+    test1, _, combined = setup_combined_multiple_env()
+
+    # Set the reuse mode for the environment
+    override_env = configure_reuse(reuse_mode, combined)
+    if override_env:
+        # If there is an override environment (ie. testing reuse with
+        # an external environment) update it here.
+        test1 = override_env
+
+    test1_specs_by_hash = test1.specs_by_hash
+
+    try:
+        # Add mpileaks to the combined environment
+        with combined:
+            add("mpileaks")
+            combined.unify = False
+            combined.concretize(force=True)
+        comb_specs_by_hash = combined.specs_by_hash
+
+        # create reference env with mpileaks that does not use reuse
+        # This should concretize to the default version of mpich (3.0)
+        env("create", "new")
+        ref_env = ev.read("new")
+        with ref_env:
+            add("mpileaks")
+        ref_env.concretize()
+        ref_specs_by_hash = ref_env.specs_by_hash
+
+        # Ensure that the mpich used by the mpileaks is the mpich from the reused test environment
+        comb_mpileaks_spec = [s for s in comb_specs_by_hash.values() if s.name == "mpileaks"]
+        test1_mpich_spec = [s for s in test1_specs_by_hash.values() if s.name == "mpich"]
+        assert len(comb_mpileaks_spec) == 1
+        assert len(test1_mpich_spec) == 1
+        assert comb_mpileaks_spec[0]["mpich"].dag_hash() == test1_mpich_spec[0].dag_hash()
+
+        # None of the references specs (using mpich@3) reuse specs from test1.
+        # This tests that the reuse is not happening coincidently
+        assert not any([s in test1_specs_by_hash for s in ref_specs_by_hash])
+
+        # Changing the version of MPICH to 3 in the test env should not change
+        # the combined env.
+        with test1:
+            config("change", "packages:mpich:require:@3")
+            concretize("-f")
+        test1_specs_by_hash_updated = test1.specs_by_hash
+
+        # Concretized combinded again (this should still use the old test1 specs)
+        combined.concretize()
+        comb_specs_by_hash = combined.specs_by_hash
+
+        # Ensure that the mpich used by the mpileaks is the mpich from the originally reused test
+        # environment after updating the included test1 environment.
+        comb_mpileaks_spec = [s for s in comb_specs_by_hash.values() if s.name == "mpileaks"]
+        test1_mpich_spec = [s for s in test1_specs_by_hash.values() if s.name == "mpich"]
+        assert len(comb_mpileaks_spec) == 1
+        assert len(test1_mpich_spec) == 1
+        assert comb_mpileaks_spec[0]["mpich"].dag_hash() == test1_mpich_spec[0].dag_hash()
+
+        # None of the new test1 specs (mpich@3) should exist in the combined environment
+        assert not any([s in comb_specs_by_hash for s in test1_specs_by_hash_updated])
```

Comment:
I think we can remove this -- this is just testing that when we concretize with no new specs, we don't do anything new. That's already tested by plenty of the other concretization tests and we can simplify this test by omitting it.

If we wanted to test something different with this, we could do the reconcretization of `test1` after creating `combined` but before concretizing. Then we could test that we got the new test1 specs when `from_environment_external_test` but not on the other test cases.

Reply:
That doesn't seem to work since we re-read the included concrete at concretization time. If I use `-f` it will always pick the new MPICH, which is not what I wanted to test here.

I think this is related to additional functionality I wanted to keep old included specs, clear everything else from the environment. We can follow up for that, I don't need it for the common stacks and I think this is sufficient for most use cases.